{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# CCDI Hub Extract Samples from Reference Files\n",
    "\n",
    "This notebook cleans the Excel file from C3DC Hub and extracts samples from the MCI and RMS studies reference files from both CCDI and C3DC Hub\n",
    "\n",
    "Author: Lucy Han, Booz Allen Hamilton"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# import pandas and os\n",
    "import pandas as pd\n",
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/sbgenomics/workspace\n"
     ]
    }
   ],
   "source": [
    "# checking current path for Jupyter Notebook\n",
    "current_path = os.getcwd()\n",
    "print(current_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# src path\n",
    "src = \"/sbgenomics/project-files/Reference_Files\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# output path\n",
    "output = \"/sbgenomics/output-files\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# confirm directory\n",
    "os.chdir(src)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " C3DC_phs000720_CCDI_RMS-Mut-Pred_Study_Manifest_v2.0.0.xlsx\n",
      "'C3DC_phs002790_CCDI MCI_Study_Manifest_v2.1.0.xlsx'\n",
      "'CCDI Hub_CCDI MCI_RMS Pts_Participants Download 2025-08-25 12-50-30.csv'\n",
      "'CCDI Hub_CCDI MCI_RMS_Samples Download 2025-08-25 13-24-08.csv'\n",
      "'CCDI Hub_MCI Path DICOM Files Manifest_2025-09-18 21-12-26.csv'\n",
      "'CCDI Hub_RMS-Mutation-Prediction_phs000720_Participants_Download 2025-07-18 16-11-54.csv'\n",
      "'CCDI Hub_RMS-Mutation-Prediction_Samples Download 2025-08-25 16-10-57.csv'\n",
      "'CCDI Hub_RMS Path DICOM Files Manifest_2025-09-18 21-24-21.csv'\n",
      " Filtered\n",
      " img_rand_sample_phs002790_phs000720.csv\n"
     ]
    }
   ],
   "source": [
    "# list the files in the directory\n",
    "!ls"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Read in reference files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# MCI Reference File names\n",
    "MCI_study_file = 'C3DC_phs002790_CCDI MCI_Study_Manifest_v2.1.0.xlsx'\n",
    "MCI_part_file = 'CCDI Hub_CCDI MCI_RMS Pts_Participants Download 2025-08-25 12-50-30.csv'\n",
    "MCI_samples_file = 'CCDI Hub_CCDI MCI_RMS_Samples Download 2025-08-25 13-24-08.csv'\n",
    "MCI_dicom_file = 'CCDI Hub_MCI Path DICOM Files Manifest_2025-09-18 21-12-26.csv'\n",
    "\n",
    "# RMS Reference File names\n",
    "RMS_study_file = 'C3DC_phs000720_CCDI_RMS-Mut-Pred_Study_Manifest_v2.0.0.xlsx'\n",
    "RMS_part_file = 'CCDI Hub_RMS-Mutation-Prediction_phs000720_Participants_Download 2025-07-18 16-11-54.csv'\n",
    "RMS_samples_file = 'CCDI Hub_RMS-Mutation-Prediction_Samples Download 2025-08-25 16-10-57.csv'\n",
    "RMS_dicom_file = 'CCDI Hub_RMS Path DICOM Files Manifest_2025-09-18 21-24-21.csv'\n",
    "\n",
    "# Image sample file (sample was generated outside of the SB-CGC environment)\n",
    "image_sample_file = 'img_rand_sample_phs002790_phs000720.csv'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read MCI files\n",
    "MCI_part_ref_df = pd.read_csv(MCI_part_file)\n",
    "MCI_samples_ref_df = pd.read_csv(MCI_samples_file)\n",
    "MCI_dicom_ref_df = pd.read_csv(MCI_dicom_file)\n",
    "\n",
    "# Read RMS files\n",
    "RMS_part_ref_df = pd.read_csv(RMS_part_file)\n",
    "RMS_samples_ref_df = pd.read_csv(RMS_samples_file)\n",
    "RMS_dicom_ref_df = pd.read_csv(RMS_dicom_file)\n",
    "\n",
    "# read image sample\n",
    "sample_ref_df = pd.read_csv(image_sample_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# check shape of the dataframe\n",
    "sample_ref_df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read MCI Excel study file\n",
    "MCI_xl = pd.ExcelFile(MCI_study_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read RMS Excel study file\n",
    "RMS_xl = pd.ExcelFile(RMS_study_file)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Process Excel files from C3DC Hub\n",
    "\n",
    "The Excel files have many tabs. This section filters out the tabs in the Excel file that do not have a participant.participant_id column and returns a new Excel sheet with only the tabs that have participant.participant_id that have been selected in the sample."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "def process_excel_files(input_xl_file):\n",
    "    # function to process excel files\n",
    "    \n",
    "    # input_xl_file is the file path to the excel file to read in\n",
    "    \n",
    "    input_excel_file = input_xl_file # input excel file path\n",
    "    output_excel_file = f\"{output}/filtered_{input_xl_file}\" # Output excel file path\n",
    "    ref_identifier_column_name = 'Participant ID'  # Column name for identifiers in both the DataFrame and Excel sheets\n",
    "    identifier_column_name = 'participant.participant_id'\n",
    "    \n",
    "    identifier_list = sample_ref_df[ref_identifier_column_name].astype(str).tolist()\n",
    "    \n",
    "    # load excel file\n",
    "    xlsx = pd.ExcelFile(input_excel_file)\n",
    "    sheet_names = xlsx.sheet_names\n",
    "    \n",
    "    # Use ExcelWriter to outp\n",
    "    with pd.ExcelWriter(output_excel_file, engine='openpyxl') as writer:\n",
    "        for sheet in sheet_names:\n",
    "            df = pd.read_excel(input_excel_file, sheet_name=sheet)\n",
    "            \n",
    "            # Skip sheet if identifier column doesn't exist\n",
    "            if identifier_column_name not in df.columns:\n",
    "                print(f\"Skipping sheet '{sheet}' - no '{identifier_column_name}' column found.\")\n",
    "                continue\n",
    "            \n",
    "            # Ensure identifier column is string type\n",
    "            df[identifier_column_name] = df[identifier_column_name].astype(str)\n",
    "    \n",
    "            # Filter based on identifiers\n",
    "            filtered_df = df[df[identifier_column_name].isin(identifier_list)]\n",
    "    \n",
    "            # Write filtered data to the same-named sheet in the output file\n",
    "            filtered_df.to_excel(writer, sheet_name=sheet, index=False)\n",
    "    \n",
    "    print(f\"Filtered Excel saved to: {output_excel_file}\")\n",
    "    return output_excel_file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Skipping sheet 'README and INSTRUCTIONS' - no 'participant.participant_id' column found.\n",
      "Skipping sheet 'study' - no 'participant.participant_id' column found.\n",
      "Skipping sheet 'study_admin' - no 'participant.participant_id' column found.\n",
      "Skipping sheet 'study_arm' - no 'participant.participant_id' column found.\n",
      "Skipping sheet 'study_funding' - no 'participant.participant_id' column found.\n",
      "Skipping sheet 'study_personnel' - no 'participant.participant_id' column found.\n",
      "Skipping sheet 'publication' - no 'participant.participant_id' column found.\n",
      "Skipping sheet 'participant' - no 'participant.participant_id' column found.\n",
      "Skipping sheet 'cell_line' - no 'participant.participant_id' column found.\n",
      "Skipping sheet 'pdx' - no 'participant.participant_id' column found.\n",
      "Skipping sheet 'sequencing_file' - no 'participant.participant_id' column found.\n",
      "Skipping sheet 'methylation_array_file' - no 'participant.participant_id' column found.\n",
      "Skipping sheet 'cytogenomic_file' - no 'participant.participant_id' column found.\n",
      "Skipping sheet 'pathology_file' - no 'participant.participant_id' column found.\n",
      "Skipping sheet 'Dictionary' - no 'participant.participant_id' column found.\n",
      "Skipping sheet 'Terms and Value Sets' - no 'participant.participant_id' column found.\n",
      "Filtered Excel saved to: /sbgenomics/output-files/filtered_C3DC_phs002790_CCDI MCI_Study_Manifest_v2.1.0.xlsx\n"
     ]
    }
   ],
   "source": [
    "filtered_MCI_study_file = process_excel_files(MCI_study_file)\n",
    "\n",
    "filtered_MCI_study_xl = pd.ExcelFile(filtered_MCI_study_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['diagnosis',\n",
       " 'survival',\n",
       " 'treatment',\n",
       " 'treatment_response',\n",
       " 'synonym',\n",
       " 'family_relationship',\n",
       " 'medical_history',\n",
       " 'exposure',\n",
       " 'radiology_file',\n",
       " 'molecular_test',\n",
       " 'sample_RMS',\n",
       " 'sample',\n",
       " 'clinical_measure_file',\n",
       " 'generic_file']"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "filtered_MCI_study_xl.sheet_names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Skipping sheet 'README and INSTRUCTIONS' - no 'participant.participant_id' column found.\n",
      "Skipping sheet 'study' - no 'participant.participant_id' column found.\n",
      "Skipping sheet 'study_admin' - no 'participant.participant_id' column found.\n",
      "Skipping sheet 'study_arm' - no 'participant.participant_id' column found.\n",
      "Skipping sheet 'study_funding' - no 'participant.participant_id' column found.\n",
      "Skipping sheet 'study_personnel' - no 'participant.participant_id' column found.\n",
      "Skipping sheet 'publication' - no 'participant.participant_id' column found.\n",
      "Skipping sheet 'participant' - no 'participant.participant_id' column found.\n",
      "Skipping sheet 'cell_line' - no 'participant.participant_id' column found.\n",
      "Skipping sheet 'pdx' - no 'participant.participant_id' column found.\n",
      "Skipping sheet 'sequencing_file' - no 'participant.participant_id' column found.\n",
      "Skipping sheet 'methylation_array_file' - no 'participant.participant_id' column found.\n",
      "Skipping sheet 'cytogenomic_file' - no 'participant.participant_id' column found.\n",
      "Skipping sheet 'pathology_file' - no 'participant.participant_id' column found.\n",
      "Skipping sheet 'Dictionary' - no 'participant.participant_id' column found.\n",
      "Skipping sheet 'Terms and Value Sets' - no 'participant.participant_id' column found.\n",
      "Filtered Excel saved to: /sbgenomics/output-files/filtered_C3DC_phs000720_CCDI_RMS-Mut-Pred_Study_Manifest_v2.0.0.xlsx\n"
     ]
    }
   ],
   "source": [
    "filtered_RMS_study_file = process_excel_files(RMS_study_file)\n",
    "\n",
    "filtered_RMS_study_xl = pd.ExcelFile(filtered_RMS_study_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['diagnosis',\n",
       " 'survival',\n",
       " 'treatment',\n",
       " 'treatment_response',\n",
       " 'synonym',\n",
       " 'family_relationship',\n",
       " 'medical_history',\n",
       " 'exposure',\n",
       " 'radiology_file',\n",
       " 'molecular_test',\n",
       " 'sample',\n",
       " 'clinical_measure_file',\n",
       " 'generic_file']"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "filtered_RMS_study_xl.sheet_names"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Process csv files\n",
    "\n",
    "This section filters the csv files to only include data that have been selected in the sample."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "def process_csv_files(input_file, identifier_column_name = 'Participant ID'):\n",
    "    # function to process csv files\n",
    "    \n",
    "    # input_file is the file path to the csv file to read in\n",
    "    # identifier_column_name is the id_column with the default as 'Participant ID'\n",
    "    \n",
    "    id_column_name = 'Participant ID'\n",
    "    input_csv_file = input_file      # Source CSV file\n",
    "    output_csv_file = f\"{output}/filtered_{input_csv_file}\"   # Output CSV file\n",
    "    \n",
    "    # load source data\n",
    "    source_df = pd.read_csv(input_csv_file)\n",
    "    \n",
    "    # load identifiers dataframe\n",
    "    # Replace with your real identifier list or load from a file\n",
    "    identifiers_df = sample_ref_df\n",
    "    \n",
    "    # filter rows\n",
    "    # Ensure both columns are of string type for reliable matching\n",
    "    source_df[identifier_column_name] = source_df[identifier_column_name].astype(str)\n",
    "    identifiers_df[id_column_name] = identifiers_df[id_column_name].astype(str)\n",
    "    \n",
    "    # Filter source data where IDs are in identifiers list\n",
    "    filtered_df = source_df[source_df[identifier_column_name].isin(identifiers_df[id_column_name])]\n",
    "    \n",
    "    # save dataframe\n",
    "    filtered_df.to_csv(output_csv_file, index=False)\n",
    "    \n",
    "    print(f\"Filtered data saved to: {output_csv_file}\")\n",
    "    return output_csv_file"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Process MCI CSV Files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtered data saved to: /sbgenomics/output-files/filtered_CCDI Hub_CCDI MCI_RMS Pts_Participants Download 2025-08-25 12-50-30.csv\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "(71, 9)"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "filtered_MCI_part_file = process_csv_files(MCI_part_file)\n",
    "filtered_MCI_part_df = pd.read_csv(filtered_MCI_part_file)\n",
    "\n",
    "filtered_MCI_part_df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtered data saved to: /sbgenomics/output-files/filtered_CCDI Hub_CCDI MCI_RMS_Samples Download 2025-08-25 13-24-08.csv\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "(163, 8)"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "filtered_MCI_samples_file = process_csv_files(MCI_samples_file)\n",
    "filtered_MCI_samples_df = pd.read_csv(filtered_MCI_samples_file)\n",
    "\n",
    "filtered_MCI_samples_df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtered data saved to: /sbgenomics/output-files/filtered_CCDI Hub_MCI Path DICOM Files Manifest_2025-09-18 21-12-26.csv\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "(462, 10)"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "filtered_MCI_dicom_file = process_csv_files(MCI_dicom_file)\n",
    "filtered_MCI_dicom_df = pd.read_csv(filtered_MCI_dicom_file)\n",
    "\n",
    "filtered_MCI_dicom_df.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Process RMS CSV Files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtered data saved to: /sbgenomics/output-files/filtered_CCDI Hub_RMS-Mutation-Prediction_phs000720_Participants_Download 2025-07-18 16-11-54.csv\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "(34, 9)"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "filtered_RMS_part_file = process_csv_files(RMS_part_file)\n",
    "filtered_RMS_part_df = pd.read_csv(filtered_RMS_part_file)\n",
    "\n",
    "filtered_RMS_part_df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtered data saved to: /sbgenomics/output-files/filtered_CCDI Hub_RMS-Mutation-Prediction_Samples Download 2025-08-25 16-10-57.csv\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "(34, 8)"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "filtered_RMS_samples_file = process_csv_files(RMS_samples_file)\n",
    "filtered_RMS_samples_df = pd.read_csv(filtered_RMS_samples_file)\n",
    "\n",
    "filtered_RMS_samples_df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtered data saved to: /sbgenomics/output-files/filtered_CCDI Hub_RMS Path DICOM Files Manifest_2025-09-18 21-24-21.csv\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "(246, 9)"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "filtered_RMS_dicom_file = process_csv_files(RMS_dicom_file, identifier_column_name = 'Sample ID')\n",
    "filtered_RMS_dicom_df = pd.read_csv(filtered_RMS_dicom_file)\n",
    "\n",
    "filtered_RMS_dicom_df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " C3DC_phs000720_CCDI_RMS-Mut-Pred_Study_Manifest_v2.0.0.xlsx\n",
      "'C3DC_phs002790_CCDI MCI_Study_Manifest_v2.1.0.xlsx'\n",
      "'CCDI Hub_CCDI MCI_RMS Pts_Participants Download 2025-08-25 12-50-30.csv'\n",
      "'CCDI Hub_CCDI MCI_RMS_Samples Download 2025-08-25 13-24-08.csv'\n",
      "'CCDI Hub_MCI Path DICOM Files Manifest_2025-09-18 21-12-26.csv'\n",
      "'CCDI Hub_RMS-Mutation-Prediction_phs000720_Participants_Download 2025-07-18 16-11-54.csv'\n",
      "'CCDI Hub_RMS-Mutation-Prediction_Samples Download 2025-08-25 16-10-57.csv'\n",
      "'CCDI Hub_RMS Path DICOM Files Manifest_2025-09-18 21-24-21.csv'\n",
      " img_rand_sample_phs002790_phs000720.csv\n"
     ]
    }
   ],
   "source": [
    "!ls"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
